src/com/amazon/corretto/crypto/provider/HpkeCipher.java

```diff
@@ -0,0 +1,309 @@
+  private EvpHpkeKey key_;
+  private HpkeParameterSpec params_;
+
+  private final Object lock_ = new Object();
+
+  HpkeCipher(AmazonCorrettoCryptoProvider provider) {
+    Loader.checkNativeLibraryAvailability();
+    provider_ = provider;
+  }
+
+  // Core Native Methods
+  // -------------------
+
+  /**
+   * Performs single-shot HPKE encryption or decryption, as specified in Section 6.1 of RFC 9180.
+   *
+   * @return number of bytes written to output, and -1 if failed.
+   */
+  private static native int hpkeCipher(
+      long keyHandle,
+      int javaCipherMode,
+      int kemId,
+      int KdfId,
+      int aeadId,
+      byte[] input,
+      int inputOffset,
+      int inputLen,
+      byte[] output,
+      int outputOffset);
+
+  /**
+   * Computes the number of bytes the output buffer needs to be for wrapping and unwrapping.
+   *
+   * <p>For wrapping, the size is greater than the input buffer since the output buffer also needs
+   * to include an AEAD tag and KEM encapsulate.
+   *
+   * <p>For unwrapping, the size is smaller, since the output does not need the AEAD tag or the KEM
+   * encapsulate.
+   */
+  private static native int hpkeOutputSize(
+      int javaCipherMode, int kemId, int KdfId, int aeadId, int inputLen);
+
+  // Core Java Methods
+  // -----------------
+
+  @Override
+  protected void engineInit(int opmode, Key key, AlgorithmParameterSpec params, SecureRandom random)
+      throws InvalidKeyException, InvalidAlgorithmParameterException {
+    if ((opmode != Cipher.WRAP_MODE) && (opmode != Cipher.UNWRAP_MODE)) {
+      throw new IllegalStateException("HpkeCipher only supports WRAP_MODE and UNWRAP_MODE");
+    }
+    if (params == null) {
+      throw new InvalidAlgorithmParameterException(
+          "HpkeCipher does not support a null parameters.");
+    }
+    if (!(params instanceof HpkeParameterSpec)) {
+      throw new InvalidAlgorithmParameterException(
+          "HpkeCipher only supports HpkeParameterSpec parameters.");
+    }
+    if (!(key instanceof EvpHpkeKey)) {
+      throw new InvalidKeyException("HpkeCipher only supports EvpHpkeKey keys.");
+    }
+    if (((EvpHpkeKey) key).getSpec() != params) {
+      throw new InvalidKeyException("Spec of key does not match the params provided");
+    }
+    if ((opmode == Cipher.WRAP_MODE) && !(key instanceof EvpHpkePublicKey)) {
+      throw new IllegalStateException("Need PublicKey to wrap");
+    }
+    if ((opmode == Cipher.UNWRAP_MODE) && !(key instanceof EvpHpkePrivateKey)) {
+      throw new IllegalStateException("Need PrivateKey to unwrap");
+    }
+    synchronized (lock_) {
+      params_ = (HpkeParameterSpec) params;
+      javaCipherMode_ = opmode;
+      key_ = opmode == Cipher.WRAP_MODE ? (EvpHpkePublicKey) key : (EvpHpkePrivateKey) key;
+    }
+  }
+
+  /**
+   * Performs HPKE single-shot encryption, as specified in Section 6.1 of RFC 9180.
+   *
+   * @return concatenation of KEM encapsulated key and encrypted ciphertext
```

Comment:
Is this concatenation format standardized or is it unique to your implementation? AFAICT, RFC 9180 simply refers to the encapsulated key and ciphertext as separate things. If it's unique to this implementation this should be documented more clearly somewhere and minimally the mirror logic should be described during unwrap.

Reply:
This pattern of concatenating the encapsulated key and the ciphertext is a  well-know, see for instance: https://libsodium.gitbook.io/doc/public-key_cryptography/sealed_boxes#algorithm-details

I updated the javadocs to be more clear